src/charm.py

```diff
@@ -31,12 +31,53 @@ def __init__(self, *args):
 
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.mongodb_relation_joined, self._mongodb_relation_joined)
         self.framework.observe(self.on.start, self._on_start)
 
+    def _mongodb_relation_joined(self, event) -> None:
+        if not self.unit.is_leader():
+            return
+
+        # do not initialise replica set until all peers have joined
+        if not self.app.planned_units() == len(self._peers.units) + 1:
+            logger.debug(
+                "planned units does not match joined units: %s != %s, will "
+                + "wait to initialise replica set until all planned units "
+                + "have joined",
+                str(self.app.planned_units()),
+                len(self.model.get_relation(PEER).units) + 1,
+            )
+            return
+
+        # do not initilise replica set until all peers have started mongod
+        for peer in self._peers.units:
+            peer_ip = str(self._peers.data[peer].get("private-address"))
+            peer_config = self._config
+            peer_config["calling_unit_ip"] = peer_ip
+            mongo_peer = MongoDB(peer_config)
+            if not mongo_peer.is_ready(all_replicas=False):
+                logger.debug(
+                    "unit: %s is not ready, cannot initilise replica set "
+                    + "until all units are ready, defering on relation-joined",
+                    peer,
+                )
+                event.defer()
+                return
+
+        # initialise replica set if not already done
+        if not self._mongo.is_replica_set():
+            self._initialise_replica_set()
+
+        # verify replica set is ready
+        if not self._mongo.is_ready():
+            logger.debug("Replica set for units: %s, is not ready", self._unit_ips)
+            self.unit.status = BlockedStatus("failed to initialise replica set")
```

Comment:
Yes, so the `is_ready` here checks if the whole replicaset after initialisation is ready but the one above checks if each individual unit of MongoDB as the mongod (MongoDDB dameon running) I think I should change the optional variable name here so it makes more sense, let me tag you in that conversation. :)

Reply:
Sorry Mia. I probably confused you with the word `above`. The call to `is ready` that I intended to comment about was the one from L72. The question is related to the fact that the `BlockedStatus` that is set inside the `if` clause is also set in L228, which is called in L69. I believe that there are two calls to set the same status. Maybe it's not the case if we can have `is ready` returning `False` after L224 not throwing an error. We can clarify this point better on MM.

Reply:
Hmm I see what you're saying 🤔 . 

It's hard to say, I did this because I thought it _might_  be possible that `self._initialise_replica_set()` was successful but for some unforeseen reason `_mongo` might not be ready. `self._initialise_replica_set()` doesn't check if the replicaset as a whole is ready and the checks for `is_ready()` above only check if each unit has `mongod` running not the whole replicaset. So I think this check is still needed.

It might be best if we get the opinion of someone that knows MongoDB a bit better here. @balbirthomas what do you think of this?

Reply:
Hi Mia! Thanks for the explanation and the details in our meeting today. I understood that makes sense to have this additional check for some situations where there is some time needed to the entire replica set to be ready.